File: frontend/src/types/alerta.ts
// Tipos de alertas del sistema
export type TipoAlerta =
  | 'vacuna'
  | 'herraje'
  | 'pago'
  | 'evento'
  | 'cumpleaños'
  | 'contrato'
  | 'stock'
  | 'tarea'
  | 'mantenimiento'
  | 'veterinaria'
  | 'otro';

export type Prioridad = 'baja' | 'media' | 'alta' | 'critica';

export type FrecuenciaAlerta =
  | 'unica'
  | 'diaria'
  | 'semanal'
  | 'mensual'
  | 'cada_x_dias';

// Alerta individual
export interface Alerta {
  id: string;
  tipo_alerta_id?: string; // Referencia al tipo de alerta configurado
  usuario_id?: string;
  tipo: TipoAlerta;
  prioridad: Prioridad;
  titulo: string;
  mensaje: string;
  leida: boolean;
  fecha_evento?: string; // Fecha del evento que genera la alerta
  fecha_vencimiento?: string; // Cuándo expira la alerta
  entidad_relacionada_tipo?: string; // 'caballo', 'cliente', 'pago', etc.
  entidad_relacionada_id?: string;
  acciones_disponibles?: AccionAlerta[]; // Acciones que se pueden realizar desde la alerta
  datos_adicionales?: Record<string, any>; // Metadata adicional
  created_at: string;
  updated_at: string;
}

// Acción que se puede realizar desde una alerta
export interface AccionAlerta {
  tipo: 'ver_detalle' | 'marcar_completada' | 'posponer' | 'ir_a_modulo' | 'custom';
  etiqueta: string;
  url?: string;
  icono?: string;
}

// Configuración de tipo de alerta (plantilla)
export interface TipoAlertaConfig {
  id: string;
  nombre: string;
  tipo: TipoAlerta;
  descripcion?: string;
  activo: boolean;
  prioridad_default: Prioridad;

  // Frecuencia y repetición
  frecuencia: FrecuenciaAlerta;
  dias_anticipacion?: number; // Días antes del evento para enviar la alerta
  intervalo_dias?: number; // Para frecuencia 'cada_x_dias'
  hora_envio?: string; // HH:MM formato 24h

  // Destinatarios
  enviar_a_roles?: string[]; // Roles que reciben esta alerta
  enviar_a_usuarios?: string[]; // IDs de usuarios específicos
  enviar_a_responsables?: boolean; // Si se envía al responsable de la entidad

  // Canales de envío
  canal_sistema: boolean; // Aparece en el sistema (campana)
  canal_email: boolean;
  canal_push?: boolean;

  // Plantilla de mensaje
  plantilla_titulo?: string; // Puede tener variables como {nombre}, {fecha}, etc.
  plantilla_mensaje?: string;

  // Condiciones de activación
  condiciones?: CondicionAlerta[];

  created_at: string;
  updated_at: string;
}

// Condición para activar una alerta
export interface CondicionAlerta {
  campo: string; // Ej: 'dias_hasta_vencimiento', 'monto_pendiente', etc.
  operador: 'igual' | 'mayor' | 'menor' | 'mayor_igual' | 'menor_igual' | 'entre' | 'contiene';
  valor: any;
  valor_maximo?: any; // Para operador 'entre'
}

// Registro de envío de alerta
export interface RegistroAlerta {
  id: string;
  alerta_id: string;
  usuario_id: string;
  canal: 'sistema' | 'email' | 'push';
  estado: 'pendiente' | 'enviado' | 'fallido' | 'leido';
  fecha_envio?: string;
  fecha_lectura?: string;
  error_mensaje?: string;
  created_at: string;
}

// Configuración de alertas del usuario
export interface ConfiguracionAlertas {
  usuario_id: string;
  alertas_sistema: boolean; // Recibir alertas en el sistema
  alertas_email: boolean;
  alertas_push: boolean;

  // Preferencias por tipo
  tipos_alertas: {
    vacuna: boolean;
    herraje: boolean;
    pago: boolean;
    evento: boolean;
    cumpleaños: boolean;
    contrato: boolean;
    stock: boolean;
    tarea: boolean;
    mantenimiento: boolean;
    veterinaria: boolean;
    otro: boolean;
  };

  // Horarios permitidos
  horario_inicio?: string; // HH:MM
  horario_fin?: string; // HH:MM
  dias_semana?: number[]; // 0-6, domingo es 0

  // Agrupación
  agrupar_alertas: boolean; // Enviar resumen en vez de una por una
  intervalo_agrupacion?: number; // Minutos
}

// DTOs para crear/actualizar
export interface TipoAlertaConfigCreate {
  nombre: string;
  tipo: TipoAlerta;
  descripcion?: string;
  prioridad_default: Prioridad;
  frecuencia: FrecuenciaAlerta;
  dias_anticipacion?: number;
  intervalo_dias?: number;
  hora_envio?: string;
  enviar_a_roles?: string[];
  enviar_a_usuarios?: string[];
  enviar_a_responsables?: boolean;
  canal_sistema: boolean;
  canal_email: boolean;
  canal_push?: boolean;
  plantilla_titulo?: string;
  plantilla_mensaje?: string;
  condiciones?: CondicionAlerta[];
}

export interface TipoAlertaConfigUpdate {
  nombre?: string;
  descripcion?: string;
  activo?: boolean;
  prioridad_default?: Prioridad;
  frecuencia?: FrecuenciaAlerta;
  dias_anticipacion?: number;
  intervalo_dias?: number;
  hora_envio?: string;
  enviar_a_roles?: string[];
  enviar_a_usuarios?: string[];
  enviar_a_responsables?: boolean;
  canal_sistema?: boolean;
  canal_email?: boolean;
  canal_push?: boolean;
  plantilla_titulo?: string;
  plantilla_mensaje?: string;
  condiciones?: CondicionAlerta[];
}

export interface AlertaCreate {
  tipo_alerta_id?: string;
  usuario_id?: string;
  tipo: TipoAlerta;
  prioridad: Prioridad;
  titulo: string;
  mensaje: string;
  fecha_evento?: string;
  fecha_vencimiento?: string;
  entidad_relacionada_tipo?: string;
  entidad_relacionada_id?: string;
  acciones_disponibles?: AccionAlerta[];
  datos_adicionales?: Record<string, any>;
}

// Estadísticas de alertas
export interface EstadisticasAlertas {
  total_alertas: number;
  alertas_no_leidas: number;
  alertas_por_prioridad: Record<Prioridad, number>;
  alertas_por_tipo: Record<TipoAlerta, number>;
  alertas_vencidas: number;
  alertas_hoy: number;
  alertas_esta_semana: number;
}

// Filtros para buscar alertas
export interface FiltrosAlertas {
  tipo?: TipoAlerta;
  prioridad?: Prioridad;
  leida?: boolean;
  fecha_desde?: string;
  fecha_hasta?: string;
  usuario_id?: string;
  tipo_alerta_id?: string;
}

File: backend/app/schemas/alerta.py
from pydantic import BaseModel, Field, UUID4
from typing import Optional, List, Dict, Any
from datetime import datetime
from app.models.alerta import TipoAlertaEnum, PrioridadAlertaEnum, FrecuenciaAlertaEnum


# ========== ALERTAS ==========

class AlertaBase(BaseModel):
    """Schema base de Alerta"""
    tipo: TipoAlertaEnum
    prioridad: PrioridadAlertaEnum = PrioridadAlertaEnum.MEDIA
    titulo: str = Field(..., min_length=1, max_length=255)
    mensaje: str


class AlertaCreate(AlertaBase):
    """Schema para crear Alerta"""
    tipo_alerta_id: Optional[UUID4] = None
    usuario_id: Optional[UUID4] = None
    fecha_evento: Optional[datetime] = None
    fecha_vencimiento: Optional[datetime] = None
    entidad_relacionada_tipo: Optional[str] = Field(None, max_length=50)
    entidad_relacionada_id: Optional[UUID4] = None
    acciones_disponibles: Optional[List[Dict[str, Any]]] = None
    datos_adicionales: Optional[Dict[str, Any]] = None


class AlertaUpdate(BaseModel):
    """Schema para actualizar Alerta"""
    leida: Optional[bool] = None
    fecha_vencimiento: Optional[datetime] = None


class AlertaSchema(AlertaBase):
    """Schema de respuesta de Alerta"""
    id: UUID4
    tipo_alerta_id: Optional[UUID4] = None
    usuario_id: Optional[UUID4] = None
    leida: bool
    fecha_evento: Optional[datetime] = None
    fecha_vencimiento: Optional[datetime] = None
    entidad_relacionada_tipo: Optional[str] = None
    entidad_relacionada_id: Optional[UUID4] = None
    acciones_disponibles: Optional[List[Dict[str, Any]]] = None
    datos_adicionales: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ========== TIPOS DE ALERTA ==========

class TipoAlertaConfigBase(BaseModel):
    """Schema base de TipoAlertaConfig"""
    nombre: str = Field(..., max_length=255)
    tipo: TipoAlertaEnum
    descripcion: Optional[str] = None
    prioridad_default: PrioridadAlertaEnum = PrioridadAlertaEnum.MEDIA
    frecuencia: FrecuenciaAlertaEnum = FrecuenciaAlertaEnum.UNICA


class TipoAlertaConfigCreate(TipoAlertaConfigBase):
    """Schema para crear TipoAlertaConfig"""
    dias_anticipacion: Optional[int] = None
    intervalo_dias: Optional[int] = None
    hora_envio: Optional[str] = None
    enviar_a_roles: Optional[List[str]] = None
    enviar_a_usuarios: Optional[List[UUID4]] = None
    enviar_a_responsables: bool = False
    canal_sistema: bool = True
    canal_email: bool = False
    canal_push: bool = False
    plantilla_titulo: Optional[str] = None
    plantilla_mensaje: Optional[str] = None
    condiciones: Optional[List[Dict[str, Any]]] = None


class TipoAlertaConfigUpdate(BaseModel):
    """Schema para actualizar TipoAlertaConfig"""
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    activo: Optional[bool] = None
    prioridad_default: Optional[PrioridadAlertaEnum] = None
    frecuencia: Optional[FrecuenciaAlertaEnum] = None
    dias_anticipacion: Optional[int] = None
    intervalo_dias: Optional[int] = None
    hora_envio: Optional[str] = None
    enviar_a_roles: Optional[List[str]] = None
    enviar_a_usuarios: Optional[List[UUID4]] = None
    enviar_a_responsables: Optional[bool] = None
    canal_sistema: Optional[bool] = None
    canal_email: Optional[bool] = None
    canal_push: Optional[bool] = None
    plantilla_titulo: Optional[str] = None
    plantilla_mensaje: Optional[str] = None
    condiciones: Optional[List[Dict[str, Any]]] = None


class TipoAlertaConfigSchema(TipoAlertaConfigBase):
    """Schema de respuesta de TipoAlertaConfig"""
    id: UUID4
    activo: bool
    dias_anticipacion: Optional[int] = None
    intervalo_dias: Optional[int] = None
    hora_envio: Optional[str] = None
    enviar_a_roles: Optional[List[str]] = None
    enviar_a_usuarios: Optional[List[UUID4]] = None
    enviar_a_responsables: bool
    canal_sistema: bool
    canal_email: bool
    canal_push: bool
    plantilla_titulo: Optional[str] = None
    plantilla_mensaje: Optional[str] = None
    condiciones: Optional[List[Dict[str, Any]]] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ========== ESTADÍSTICAS ==========

class EstadisticasAlertas(BaseModel):
    """Estadísticas de alertas"""
    total_alertas: int
    alertas_no_leidas: int
    alertas_por_prioridad: Dict[str, int]
    alertas_por_tipo: Dict[str, int]
    alertas_vencidas: int
    alertas_hoy: int
    alertas_esta_semana: int


# ========== CONFIGURACIÓN USUARIO ==========

class ConfiguracionAlertasBase(BaseModel):
    """Schema base de ConfiguracionAlertasUsuario"""
    alertas_sistema: bool = True
    alertas_email: bool = False
    alertas_push: bool = False
    tipos_alertas: Optional[Dict[str, bool]] = None
    horario_inicio: Optional[str] = None
    horario_fin: Optional[str] = None
    dias_semana: Optional[List[int]] = None
    agrupar_alertas: bool = False
    intervalo_agrupacion: Optional[int] = None


class ConfiguracionAlertasSchema(ConfiguracionAlertasBase):
    """Schema de respuesta de ConfiguracionAlertasUsuario"""
    id: UUID4
    usuario_id: UUID4
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
